lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4363,6 +4369,41 @@ void ConformanceChecker::resolveValueWitnesses() {
       break;
     }
   }
+
+  // Emit a tailored diagnostic when we have witnesses to
+  // '@objc' requirements which are default implementations.
+  if (!objcRequirementsWithDefaultImpl.empty()) {
+    SourceLoc FixItLocation;
+    SourceLoc TypeLoc;
+    auto DC = Conformance->getDeclContext();
+    if (auto Extension = dyn_cast<ExtensionDecl>(DC)) {
+      FixItLocation = Extension->getBraces().Start;
+      TypeLoc = Extension->getStartLoc();
+    } else if (auto Nominal = dyn_cast<NominalTypeDecl>(DC)) {
+      FixItLocation = Nominal->getBraces().Start;
+      TypeLoc = Nominal->getStartLoc();
+    } else {
+      llvm_unreachable("Unknown adopter kind");
+    }
+
+    auto &Diags = getASTContext().Diags;
+    Diags.diagnose(TypeLoc, diag::type_does_not_conform, Conformance->getType(),
+                   Proto->getDeclaredInterfaceType());
+    Diags.diagnose(TypeLoc, diag::witness_objc_default_implementation);
+
+    if (getASTContext().LangOpts.DiagnosticsEditorMode) {
+      llvm::SmallString<128> FixItString;
+      llvm::raw_svector_ostream FixItStream(FixItString);
+      for (auto requirement : objcRequirementsWithDefaultImpl) {
+        printRequirementStub(requirement, Conformance->getDeclContext(),
```

Comment:
If there are other unsatisfied requirements that aren't implemented in an extension, will we print stubs for those too? I'm also not too familiar with the conformance checker implementation - can we unify this with the code that already diagnoses missing witnesses?

Reply:
Missing witnesses will be handled by existing code, so we don't need to check for that here. The difference here is that the witness isn't missing (the default implementation is the witness), so I am not sure if it makes sense to unify the code with the "missing witnesses" code elsewhere, but if this isn't the right place to do this check then I would appreciate some suggestions on where to move it to!

Reply:
What happens if the user writes something like:

```swift
import Foundation 

@objc protocol ObjCProtocol {
   @objc func requiredMethod()
   @objc var requiredProperty: Int { get }
 }

 extension ObjCProtocol {
   func requiredMethod() {}
 }

 class ConformingClass: NSObject, ObjCProtocol {} 
```

Are they going to get two error messages saying `Type 'ConformingClass' does not conform to protocol 'ObjCProtocol'` with two fix-its to insert stubs?

Reply:
Yes! One of them with have a note attached to it to say that default implementations are not allowed. I guess that's not ideal...

Reply:
Here’s an idea (may or may not be good) - what if we didn’t consider the default implementation as a legitimate witness in this case? Then we’ll end up in the “missing witnesses” path which can deal with stubs and etc. We can just emit the tailored note here.

Reply:
That sounds like a good idea to me. @slavapestov is there any reason the above suggestion would be incorrect? 

Also, just for my understanding, is there any point in implementing an `@objc` requirement of an `@objc` protocol in a protocol extension? Could you still call it like `ObjCProtocol.requiredMethod(ConformingClass())()`?

Reply:
Friendly ping, @slavapestov